src/bookshelf/src/bksw_sql.erl

```diff
@@ -0,0 +1,253 @@
+%% @copyright 2015 Chef, Inc. All Rights Reserved
+%% @author Mark Anderson <mark@chef.io>
+%%
+%% Licensed to the Apache Software Foundation (ASF) under one or more
+%% contributor license agreements.  See the NOTICE file distributed
+%% with this work for additional information regarding copyright
+%% ownership.  The ASF licenses this file to you under the Apache
+%% License, Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain a copy of
+%% the License at http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+%% implied.  See the License for the specific language governing
+%% permissions and limitations under the License.
+-module(bksw_sql).
+
+-include("internal.hrl").
+
+%-ifdef(namespaced_types). %% workaround for sqerl complaint
+-type dict() :: dict:dict().
+%-endif.
+
+-export([ statements/1,
+          ping/0,
+          create_bucket/1,
+          create_bucket_b/1,
+          delete_bucket/1,
+          delete_bucket_b/1,
+          find_bucket/1,
+          list_buckets/0,
+          list_bucket/1,
+          bucket_exists/1,
+          find_file/2,
+          create_file_link_data/3,
+          link_file_data/2,
+          delete_file/1,
+          insert_file_data/0,
+          add_file_chunk/3,
+          update_metadata/6,
+          get_chunk_data/2,
+          init_transfer_state/0,
+          update_transfer_state/3,
+          finalize_transfer_state/2
+        ]).
+
+-include_lib("sqerl/include/sqerl.hrl").
+
+ping() ->
+    case sqerl:select(ping, [], first_as_scalar, [ping]) of
+        {ok, <<"pong">>} -> ok;
+        {error, Reason} -> {error, Reason}
+    end.
+
+%% @doc bookshelf queries.
+%% Referenced by sys.config in {sqerl, [{ prepared_staements...
+statements(_) ->
+    Path = filename:join([code:priv_dir(bookshelf), "pgsql_statements.config"]),
+    {ok, Statements} = file:consult(Path),
+    Statements.
+
+create_bucket(BucketName) ->
+    case sqerl:statement(insert_bucket, [BucketName], count) of
+        {ok, 1} ->
+            ok;
+        {conflict, _} ->
+            error("Never should happen");
```

Comment:
>  What about simultaneous calls to the bucket creation API interleaving these database requests?

I think the only caller of this is create_bucket_b which checks if the bucket exists first.  `create_bucket_b` is what is called in the controller, so concurrent bucket creates should still never hit this.  That said, this could probably be refactored a bit for sure.